### ISSUE TITLE
Meteor will Self-Terminate Under High CPU Use and/or Heavy Method Call Stress

### ISSUE THREAD

**Issue by meawoppl** [meteor/meteor#1302]
TL;DR The current proxy and sub-process management can kill otherwise well-behaved meteor projects when high CPU loads are encountered.  The characteristic behavior is seen as:

> Failed to receive keepalive! Exiting.
> => Exited with code: 1
> 
> events.js:71
> ...
> less useful stack barf

The outer proxy that Meteor sets up pings the inner running process every two seconds with the letter "k" to signify that it should not commit ritual seppuku to please its master:

``` javascript
// Keepalive so server can detect when we die
  var timer = setInterval(function () {
    try {
      if (proc && proc.pid && proc.stdin && proc.stdin.write)
        proc.stdin.write('k');
    } catch (e) {
      // do nothing. this fails when the process dies.
    }
  }, 2000);
```

On the inner server side this is met with the following function,

``` javascript
var init_keepalive = function () {
  var keepalive_count = 0;

  process.stdin.on('data', function (data) {
    keepalive_count = 0;
  });

  process.stdin.resume();

  setInterval(function () {
    keepalive_count ++;
    if (keepalive_count >= 3) {
      console.log("Failed to receive keepalive! Exiting.");
      process.exit(1);
    }
  }, 3000);
};
```

which is run contingent on a hard-wired sub-process environment argument, found on [this line](https://github.com/meteor/meteor/blob/2e78d0d110443ac6fb9471e6be1c136f15f03fd2/tools/run.js#L248)

This wiring introduces three dependencies:
 -the outer shell forwarding stuff promptly
 -the correct performance of the outer proxy
 -additionally a number of these callbacks can stack up and be executed roughly at the same time ([libev manual for the interested](http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#ANATOMY_OF_A_WATCHER)), but the high level Node.js disclaimer that, ["Node.js makes no guarantees about the exact timing of when the callback will fire, nor of the ordering things will fire in. The callback will be called as close as possible to the time specified."](http://nodejs.org/api/timers.html#timers_settimeout_callback_delay_arg), pretty much wraps that up.

Anyway, my suggestion, (will write pull request if desired) would be to re-write the inner server function with a setTimeout based keepalive poll, which re-registers itself when run.  

The fact this is passed as an environment variable suggests that this is an optional behavior and might not be used in the bundled version of apps?
